### PR TITLE
Add support for reading routing operational data for IPv4

### PR DESCRIPTION
**Pull request by mattiaswal**
In the CLI:
admin@infix-00-00-00:/> show routes
```
PREFIX                        DESTINATION                   PROTOCOL  METRIC
1.1.1.0/24                    e0                            kernel
192.168.100.0/24              1.1.1.1                       static    20
192.168.110.2/32              blackhole                     static    20
192.168.120.2/32              blackhole                     static    20
192.168.130.2/32              unreachable                   static    20
```
EDIT by @rical : framed output in markdown ```

### CODE REVIEW

File: board/netconf/rootfs/lib/infix/cli-pretty [hunk 1]
```diff
@@ -297,6 +329,24 @@ def ietf_interfaces(json, name):
...
 
+def ietf_routing(json, protocol="ipv4"):
+    if not json.get("ietf-routing:routing"):
+        print(f"Error, top level \"ietf-routing:routing\" missing")
+        sys.exit(1)
+    routes = get_json_data({}, json, 'ietf-routing:routing','ribs', 'rib')[0]
+    routes = get_json_data(None, routes, "routes", "route")
+
+    hdr = (f"{'PREFIX':<{PadRoute.prefix}}"
+           f"{'NEXT HOP':<{PadRoute.next_hop}}"
```
Comment (rical) [inline board/netconf/rootfs/lib/infix/cli-pretty:340]: Perhaps `NEXT-HOP` would look nicer in the header?

File: src/statd/statd.c [hunk 2]
```diff
@@ -173,8 +184,8 @@ static int ly_add_yanger_data(const struct ly_ctx *ctx, struct lyd_node **parent
 }
 
 static int sr_ifaces_cb(sr_session_ctx_t *session, uint32_t, const char *path,
-			const char *, const char *, uint32_t,
-			struct lyd_node **parent, void *priv)
+                        const char *, const char *, uint32_t,
```
Comment (rical) [inline src/statd/statd.c:187]: I believe this indentation was "the kernel way"?

File: src/statd/statd.c [hunk 3]
```diff
@@ -195,7 +206,38 @@ static int sr_ifaces_cb(sr_session_ctx_t *session, uint32_t, const char *path,
...
+
+static int sr_routes_cb(sr_session_ctx_t *session, uint32_t, const char *path,
+                        const char *, const char *, uint32_t,
+                        struct lyd_node **parent, __attribute__((unused)) void *priv)
+{
+	const struct ly_ctx *ctx;
+	sr_conn_ctx_t *con;
+	sr_error_t err;
+
+	ERROR("Incoming query for xpath: %s", path);
```
Comment (rical) [inline src/statd/statd.c:226]: Accidental debug left? If you really want this line I think there should be some other macro with a more suitable name

File: src/statd/statd.c [hunk 4]
```diff
@@ -195,7 +206,38 @@ static int sr_ifaces_cb(sr_session_ctx_t *session, uint32_t, const char *path,
...
+
+	con = sr_session_get_connection(session);
+	if (!con) {
+		ERROR("Error, getting connection");
```
Comment (rical) [inline src/statd/statd.c:230]: I'm not sure if the function name is printed by the macro, but you could add "sr" or "sysrepo" in all these errors to make the message clearer to the reader. Such as:
`ERROR("Error, getting sr connection");`.

File: src/statd/statd.c [hunk 5]
```diff
@@ -221,56 +263,46 @@ static void sr_event_cb(struct ev_loop *, struct ev_io *w, int)
 	sr_subscription_process_events(sub->sr_sub, NULL, NULL);
 }
 
-static int sub_to_iface(struct statd *statd, const char *ifname)
+static int subscribe(struct statd *statd, char *model, char *xpath, const char *name,
+                     int (*cb)(sr_session_ctx_t *session, uint32_t, const char *,
+                               const char *, const char *, uint32_t,
```
Comment (rical) [inline src/statd/statd.c:268]: Indentation problem.
I believe the kernel way is
```
int long_function_nam(int foo,
<tab><tab><tab><space>int bar)
```
Where tabs are used first and if the second line doesn't line up with the bracket it's padded with spaces.

File: src/statd/statd.c [hunk 6]
```diff
@@ -221,56 +263,46 @@ static void sr_event_cb(struct ev_loop *, struct ev_io *w, int)
...
-	 * NOTE: this is a good solution but it might be to slow if a lot of
-	 * new interfaces pops up at the same time. We don't want to delay
-	 * processing the netlink messages to much.
-	 */
-	if (ip_link_check_group(ifname, "internal") == 1)
-		return SR_ERR_OK;
+	sr_error_t err;
 
-	sub = sub_find_iface(&statd->subs, ifname);
+	sub = sub_find(&statd->subs, name);
```
Comment (rical) [inline src/statd/statd.c:275]: What will happen if someone names there interface "routes"?

File: src/statd/statd.c [hunk 7]
```diff
@@ -221,56 +263,46 @@ static void sr_event_cb(struct ev_loop *, struct ev_io *w, int)
...
+	snprintf(sub->name, sizeof(sub->name), "%s", name);
 
-	DEBUG("Subscribe to events for \"%s\"", path);
-	err = sr_oper_get_subscribe(statd->sr_ses, "ietf-interfaces",
-				    path, sr_ifaces_cb, sub,
-				    SR_SUBSCR_DEFAULT | SR_SUBSCR_NO_THREAD | SR_SUBSCR_DONE_ONLY,
-				    &sub->sr_sub);
+	DEBUG("Subscribe to events for \"%s\"", xpath);
+	err = sr_oper_get_subscribe(statd->sr_ses, model,
+	                            xpath, cb, sub,
```
Comment (rical) [inline src/statd/statd.c:291]: I'm not sure about the indentation

**Review by rical — COMMENTED** [review]
Overall it looks great. I left some minor commends, mostly cosmetic. I good idea might be to run your diff though `checkpatch.pl` from the kernel.

Again, looks really good overall.

File: src/statd/statd.c (hunk 6 above)
Comment (mattiaswal) [reply src/statd/statd.c:275]: Good point, unlikely. But i fix it.

File: src/statd/statd.c (hunk 2 above)
Comment (mattiaswal) [reply src/statd/statd.c:187]: I just ran the proposed indentation by jocke made in #209, but i can see that it is spaces not tabs..

File: src/statd/statd.c (hunk 5 above)
Comment (troglobit) [reply src/statd/statd.c:268]: Indeed. I've added `src/.clang-format` from the kernel, with some minor additions at the end.  It actually works.

**Review by troglobit — CHANGES_REQUESTED** [review]
I agree with @rical's findings.

Additionally, I think I'd like to see the metric column earlier, before protocol/origin.

File: src/statd/statd.c (hunk 5 above)
Comment (mattiaswal) [reply src/statd/statd.c:268]: But it is a function pointer? Is it still valid to do so?

Comment (troglobit) [reply src/statd/statd.c:268]: Ah, of course, didn't see that at first glance. Shame on me.

**Review by troglobit — APPROVED** [review]
LGTM!

Looking forward to an update in doc/networking.md, example of how to add a route and read routing table remotely using NETCONF ;-)

File: src/statd/statd.c (hunk 5 above)
Comment (rical) [reply src/statd/statd.c]: Right, I missed that too. Sorry :)

**Pull request merged**